{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1352f623",
   "metadata": {},
   "outputs": [],
   "source": [
    "#this code for a list of genes and saves the output as a fasta file of cds or protein sequence with seq id as just the species name\n",
    "\n",
    "import requests\n",
    "import sys\n",
    "import time\n",
    "import os\n",
    "\n",
    "server = \"https://rest.ensembl.org\"\n",
    "gene_file = \"\"  # path to text file with one human gene ID per line\n",
    "output_dir = \"\"  # directory to save individual FASTA files for cds - recommend changing path when used for protein sequences\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "with open(gene_file) as f:\n",
    "    human_genes = [line.strip() for line in f if line.strip()]\n",
    "for gene in human_genes:\n",
    "    output_fasta = os.path.join(output_dir, f\"{gene}.fasta\")\n",
    "\n",
    "    # url for fetching from rest api - to fetch protein sequences change to sequence=protein \n",
    "    ext = f\"/homology/id/human/{gene}?sequence=cdna;aligned=0;type=orthologues;content-type=application/json\"\n",
    "\n",
    "    r = requests.get(server + ext, headers={\"Content-Type\": \"application/json\"})\n",
    "    if not r.ok:\n",
    "        print(f\"Error fetching {gene}\")\n",
    "        r.raise_for_status()\n",
    "        sys.exit()\n",
    "\n",
    "    decoded = r.json()\n",
    "\n",
    "    with open(output_fasta, \"w\") as out_f:\n",
    "        # for fetching human sequence id\n",
    "        human_seq = decoded.get(\"data\", [])[0].get(\"sequence\")\n",
    "        if human_seq:\n",
    "            out_f.write(f\">homo_sapiens|{gene}\\n{human_seq}\\n\")\n",
    "        else:\n",
    "            # if present inside as source - varies depending on the gene id sometimes\n",
    "            for homology in decoded.get(\"data\", [])[0].get(\"homologies\", []):\n",
    "                if homology.get(\"source\", {}).get(\"species\") == \"homo_sapiens\" :\n",
    "                    human_seq = homology[\"source\"].get(\"seq\")\n",
    "                    if human_seq:\n",
    "                        out_f.write(f\">homo_sapiens\\n{human_seq}\\n\")\n",
    "                        break\n",
    "\n",
    "        \n",
    "        for homology in decoded.get(\"data\", [])[0].get(\"homologies\", []):\n",
    "            target = homology.get(\"target\", {})\n",
    "            gene_id = target.get(\"id\")\n",
    "            species = target.get(\"species\")\n",
    "            seq = target.get(\"seq\")\n",
    "\n",
    "            \n",
    "            if species == \"human\" or species == \"homo_sapiens\":\n",
    "                continue\n",
    "\n",
    "            if gene_id and species and seq:\n",
    "                header = f\">{species}\"\n",
    "                out_f.write(f\"{header}\\n{seq}\\n\")\n",
    "\n",
    "    print(f\"Saved human + orthologs for {gene} to {output_fasta}\") #helps to keep track of successful runs\n",
    "    time.sleep(0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "819f9b07",
   "metadata": {},
   "outputs": [],
   "source": [
    "#code for filtering based on species ID - in the event of multiple sequences from same species - fetches the first one\n",
    "#use for both cds and protein sequences\n",
    "\n",
    "#pip install biopython #if not installed already\n",
    "\n",
    "from Bio import SeqIO\n",
    "import os\n",
    "\n",
    "species = set([\"homo_sapiens\", \"mus_musculus\", \"rattus_norvegicus\", \"macaca_mulatta\", \"xenopus_tropicalis\", \"sus_scrofa\"]) \n",
    "\n",
    "\n",
    "input_dir = \"\" #directory with all the orthologs fetched for a gene using the cell above\n",
    "output_dir = \"\" #output directory for storing the species level files\n",
    "input_format = \"fasta\"\n",
    "\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "for fname in os.listdir(input_dir):\n",
    "    if not fname.endswith(\".fasta\"):\n",
    "        continue\n",
    "\n",
    "    input_file = os.path.join(input_dir, fname)\n",
    "    output_file = os.path.join(output_dir, fname)\n",
    "\n",
    "    seen_species = set()\n",
    "\n",
    "    def filter_records():\n",
    "        for record in SeqIO.parse(input_file, input_format):\n",
    "            sp = record.id\n",
    "            if sp in species:\n",
    "                if sp not in seen_species:\n",
    "                    seen_species.add(sp)\n",
    "                    yield record   # only yield the first hit for this species\n",
    "\n",
    "    count = SeqIO.write(filter_records(), output_file, input_format) #to ensure that sequences from all 6 species are in the corresponding fasta files\n",
    "    print(f\"Saved {count} records from {input_file} to {output_file}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "r_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
